### ISSUE TITLE
SFT阶段很奇怪的loss震动

### ISSUE THREAD

**Issue by manfen1129** [jingyaogong/minimind#173]
![Image](https://github.com/user-attachments/assets/7d0c4552-e0f7-457a-a2d3-1cfd69d3efd6)
如图，很奇怪的loss震动。
与项目提供的SFT的loss参照不同，我的loss整体看起来挺平稳的，但是中间会出现一些巨大的波动。前者我可以解释为我的batchsize设的更大（256，是项目的一倍），但是后面的波动我确实没能理解为什么，希望可以请教一下。
具体参数如下：
LLM总参数量：25.830 百万
parser = argparse.ArgumentParser(description="MiniMind Full SFT")
parser.add_argument("--epochs", type=int, default=3)
parser.add_argument("--batch_size", type=int, default=256)
parser.add_argument("--learning_rate", type=float, default=6e-5)
parser.add_argument('--max_seq_len', default=512, type=int)
parser.add_argument("--data_path", type=str, default="./dataset/sft_512.jsonl")
设备：
单卡A800

**Comment by manfen1129** [comment]
后面的波动是否又sft_512数据集质量问题所导致？
但是吃到了糟糕的预料后，模型又快速复原了（起码loss上看没有变得更差，没有受到影响，按照趋势和前面是连续的），这感觉有点奇怪

**Comment by zxywww** [comment]
我也遭遇了这种抖动, 请问你知道是什么原因吗? 那为什么作者的loss结果没有这种抖动呢? 按理说batch_size设置的比较大, loss应该越稳定的

**Comment by manfen1129** [comment]
> 我也遭遇了这种抖动, 请问你知道是什么原因吗? 那为什么作者的loss结果没有这种抖动呢? 按理说batch_size设置的比较大, loss应该越稳定的

应该是这个数据集是拼接而成的，中间一些拼接的前后差距过大，比如有一些英语、代码部分，大batch_size一下子直接抽一大块OOD数据，导致loss之前飞了。经实验，我使用shuffle后这种抖动就消失了，你可以试一下。

**Comment by zxywww** [comment]
python train_full_sft.py --epochs 2 --batch_size 56 --learning_rate 0.00005 --use_wandb --wandb_project 'MiniMind-Full-SFT_seq_512' --num_workers 8 --dim 768 --n_layers 16 --data_path "./dataset/sft_512.jsonl"
这是我的配置参数, batch_size=56, 没有用梯度累计. loss情况如图. 应该就是数据

**Comment by zxywww** [comment]
python train_full_sft.py --epochs 2 --batch_size 56 --learning_rate 0.00005 --use_wandb --wandb_project 'MiniMind-Full-SFT_seq_512' --num_workers 8 --dim 768 --n_layers 16 --data_path "./dataset/sft_512.jsonl"
这是我的配置参数, batch_size=56, 没有用梯度累计. loss也是剧烈抖动. 应该就是数据问题. 谢谢, 我明白了, shuffle之后把这些ood数据平均分到一些batch中了, 不会聚在一起了
